### PR TITLE
[OptimizeInstructions] Simplify add / sub with negative on LHS or RHS for floating points

### PR DESCRIPTION
**Pull request by MaxGraey**
```rust
(-x) + y   ->   y - x
x + (-y)   ->   x - y
x - (-y)   ->   x + y
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed `ITERATION: 6625`

**Comment by MaxGraey** [comment]
@kripken @tlively could you take a look please?

**Review by tlively — COMMENTED** [review]
Does this definitely work for all combinations of zeros, infinities, and NaNs? LGTM if so.

**Comment by MaxGraey** [comment]
Yes, and I didn't find rule restriction in spec: https://github.com/WebAssembly/spec/blob/main/test/core/float_exprs.wast

**Comment by MaxGraey** [comment]
The only non-determinism if  `x` == `-NaN`. It will change sign of NaN. So perhaps better use this only for `fastMath`. @kripken  wdyt?

**Comment by kripken** [comment]
The sign of a NaN output of `+` and `-` is [nondeterministic](https://webassembly.github.io/spec/core/bikeshed/#nan-propagation%E2%91%A0), yeah, so it is fine to return either. So this looks valid from that PoV.

However, I am less sure whether this rule is true in IEEE 754 floats, in particular, regarding corner cases of maximum values and also the things @tlively mentioned. One way to check might be to read the IEEE 754 spec, another might be to find logic in LLVM that does the same thing.

**Comment by MaxGraey** [comment]
On LLVM I found only this: https://github.com/llvm/llvm-project/blob/09cdddea0c4d284c2c22f5dfade40a60850c5ea7/llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp#L1318

**Comment by MaxGraey** [comment]
Btw perfaps we could generalize it to any RHS value (not only for constants)

**Comment by MaxGraey** [comment]
Done. Refuzzed

File: test/lit/passes/optimize-instructions.wast [hunk 1]
```diff
@@ -10517,6 +10517,412 @@
...
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (f64.sub
+  ;; CHECK-NEXT:    (f64.neg
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (local.get $y)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $simplify-add-sub-with-neg-float (param $a f32) (param $b f32) (param $x f64) (param $y f64)
```
Comment (tlively) [inline test/lit/passes/optimize-instructions.wast:10845]: Can you break this up into multiple functions so the output is easier to see together with the input?

**Review by tlively — APPROVED** [review]

File: test/lit/passes/optimize-instructions.wast (hunk 1 above)
Comment (MaxGraey) [reply test/lit/passes/optimize-instructions.wast:10845]: Sure!

**Comment by MaxGraey** [comment]
@tlively done!

**Comment by tlively** [comment]
Emscripten errors aren't related to this PR, so going ahead and landing this.

**Comment by tlively** [comment]
Ugh, right, we generalized beyond constants but we didn't remember to handle effects properly. I'll revert this pending a fix.

**Pull request merged**